rfc/0022-policy-lifecycle.md

```diff
@@ -0,0 +1,396 @@
+This will allow us to identify the status of a policy on a specific Policy Server replica.
+
+### Policy Status
+
+We will use the [status condition](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#typical-status-properties) convention to report the status of the policy.
+The status condition struct will be extended to include the replica number.
+The approach of extending the status condition struct is already used in the Kubernetes codebase,
+for example in the [Pod conditions](https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#pod-conditions).
+
+The `PolicyStatusCondition` struct will have the following fields:
+
+| **Field name**         | **Description**                                                                                     |
+| ---------------------- | --------------------------------------------------------------------------------------------------- |
+| **type**               | Name of the policy condition.                                                                       |
+| **status**             | Indicates whether the condition is applicable, with possible values: "True", "False", or "Unknown". |
+| **reason**             | Machine-readable reason for the condition's last transition, in UpperCamelCase format.              |
+| **message**            | Human-readable message providing details about the last transition.                                 |
+| **replica** (optional) | The replica number reporting the status.                                                            |
+
+The `Policy` CRD and the `PolicyRevision` CRD will be updated to include the status conditions.
+See the [Policy and PolicyRevision conditions](#policy-and-policyrevision-conditions) section for more details on how the status conditions will be propagated.
+
+The following `PolicyConditions` will be observed:
+
+- `Scheduled`: The policy was created and is pending to be loaded.
+- `Initialized`: The policy was successfully pulled, precompiled and validated by the leader.
+- `Ready`: The policy was successfully loaded and is ready to be used. This status will include the replica number.
+
+By using the `reason` field in the status condition, we can provide detailed information in case of an error.
+For instance if the leader fails to initialize a policy because of registry error, the following status condition will be set:
+
+```yaml
+conditions:
+  - type: Initialized
+    status: False
+    reason: PullError
+    message: "Cannot pull the policy from the registry."
+```
+
+### Rollback
+
+If a policy fails to load, the Policy Server will keep running with the previous valid policy.
+The user will be able to rollback to a previous policy version by updating the `Policy` CRD (via a spec field or an annotation, TBD).
+The controller will detect the change and use the `PolicyRevision` CRD to update the Policy CRD with the previous configuration.
```

Comment:
Why do the user needs to change a version filed in the CRD? If the `Policy` CRD define the desired state, shouldn't  users just reapply the new state and see if policy server is able to load it? What happen if the user change the `Policy` CRD version field and something else (e.g. settings) ?

Reply:
I think doing a rollback is something tricky. What is described here is not what I would do.

Assuming we want to rollback from generation 2 to generation 1.
We take contents of the `PolicyRevision(g1).spec.data` and copy it back to the `Policy.spec`.

I think we could do a rollback with the set of CRDs described by this RFC. What we have to figure out is if it makes sense to offer this functionality and what the UX would be. I consider that not something to be part of this RFC